packages/api-contract/src/Abi/index.ts

```diff
@@ -181,15 +181,43 @@ export class Abi {
   }
 
   #decodeEventV5 = (record: EventRecord): DecodedEvent => {
-    const data = record.event.data[1] as Bytes;
+    // Find event by first topic, which potentially is the signature_topic
     const signatureTopic = record.topics[0];
-    const event = this.events.find((e) => e.signatureTopic !== undefined && e.signatureTopic === signatureTopic.toHex());
+    const data = record.event.data[1] as Bytes;
 
-    if (!event) {
-      throw new Error(`Unable to find event with signature_topic ${signatureTopic.toHex()}`);
+    if (signatureTopic) {
+      const event = this.events.find((e) => e.signatureTopic !== undefined && e.signatureTopic !== null && e.signatureTopic === signatureTopic.toHex());
+
+      // Early return if event found by signature topic
+      if (event) {
+        return event.fromU8a(data);
+      }
     }
 
-    return event.fromU8a(data.subarray(0));
+    // If no event returned yet, it might be anonymous
+    const amountOfTopics = record.topics.length;
+    const potentialEvents = this.events.filter((e) => {
+      // event can't have a signature topic
+      if (e.signatureTopic !== null && e.signatureTopic !== undefined) {
+        return false;
+      }
+
+      // event should have same amount of indexed fields as emitted topics
+      const amountIndexed = e.args.filter((a) => a.indexed).length;
+
+      if (amountIndexed !== amountOfTopics) {
+        return false;
+      }
+
+      // If all conditions met, it's a potential event
+      return true;
+    });
+
+    if (potentialEvents.length === 1) {
+      return potentialEvents[0].fromU8a(data);
+    }
+
+    throw new Error('Unable to determine event');
```

Comment:
The other case to consider is if a cross contract call occurs, which in turn raises an event. Since the event is raised from another contract we don't have the metadata here...so we might not want to raise an error and just give back the raw bytes instead of attempting to decode? 

This scenario might also lead to a false positive in the heuristic for above for determining anon events, if the foreign event has the same number of topics :thinking:

Reply:
Oh okay, so if contract A calls a function on B which emits an event, the event will be emitted by A and not B? 

Feels like it makes sense as users might not know anything about B so can only listen to A. 

As of now I have to little knowledge on what's happening above this little 'Abi.decodeEvent world' to have an idea how to handle this the best, yet.

Reply:
> Oh okay, so if contract A calls a function on B which emits an event, the event will be emitted by A and not B?

It depends, using `forward_call` (contract A calling contract B), the account id topic will be for contract B so would be easy to distinguish 

The other case is using `delegate_call` where the **code** of contract B is invoked from contract A, in this case the account id topic would be of contract A.